src/zip.ts

```diff
@@ -0,0 +1,42 @@
+import { purry } from './purry'
+
+/**
+ * Creates a new list from two supplied lists by pairing up equally-positioned items.
+ * The length of the returned list will match the shortest of the two inputs.
+ * @param first the first input list
+ * @param second the second input list
+ * @signature
+ *   R.zip(first, second)
+ * @example
+ *   R.zip([1, 2], ['a', 'b']) // => [1, 'a'], [2, 'b']
+ * @data_first
+ * @category Array
+ */
+export function zip<F extends Array<unknown>, S extends Array<unknown>>(first: F, second: S): Array<[F, S]>
+
+/**
+ * Creates a new list from two supplied lists by pairing up equally-positioned items.
+ * The length of the returned list will match the shortest of the two inputs.
+ * @param second the second input list
+ * @signature
+ *   R.zip(second)(first)
+ * @example
+ *   R.zip(['a', 'b'])([1, 2]) // => [[1, 'a'], [2, 'b']
+ * @data_last
+ * @category Array
+ */
+export function zip<F extends Array<unknown>, S extends Array<unknown>>(second: S): (first: F) => Array<[F, S]>
```

Comment:
I don’t think the return type of `Array<[F, S]>` is correct.
This only works for inputs that have the same length. Otherwise, they are filled up with `undefined`.

It would also be super nice to make it work with Tuples as input, so if you do:

`zip(['foo', 'bar'] as const, [1, 2] as const)`

It would yield:

`['foo', 1], ['bar', 2]` _on type level_

Reply:
Maybe we could even make zip expect an arbitrary amount of array as parameters, as shown here: https://stackoverflow.com/questions/51977209/can-a-typed-zip-function-be-implemented-now-with-typescript-3-0s-generic-rest-p

Reply:
Wow, you're right, that type's not correct, I'll revise that. 

The logic here does handle `first` and `second` being different lengths--can you provide inputs for a test case that would produce `undefined` values?

Reply:
Ah right, I mis-read the implementation. You are taking the _shorter_ length [here](https://github.com/remeda/remeda/pull/99/files#diff-3ba70535993f56e0474d4976a9e95c92R35)

So, when looking at the `lodash` implmentation, it seems that always the larger length is taken, and the missing values are filled with `undefined`:

![Screenshot 2020-10-09 at 21 37 21](https://user-images.githubusercontent.com/1021430/95624859-cfe1cd80-0a77-11eb-8b84-3a6cc2eccd83.png)

You can also see that lodash supports a variadic argument list of arrays.

Reply:
Ok, so I was working from the ramda implementation https://ramdajs.com/docs/#zip. It appears to be quite different from lodash. 

Is there a preference here of which to match when they disagree?

IMO, the `zipObj`/`zipObject` implementation of ramda/lodash is more consistent with the ramda definition of `zip` than the `lodash` definition.

Reply:
Right - i just looked up how `zip` works in other languages, e.g. [scala](https://www.scala-lang.org/api/current/scala/collection/AbstractIterable.html#zip[B](that:scala.collection.IterableOnce[B]):CC[(A@scala.annotation.unchecked.uncheckedVariance,B)]), or [haskell](https://hackage.haskell.org/package/base-4.14.0.0/docs/Data-List.html#g:17), and it seems that they all also correspond with the `ramda` logic:
- only allow two lists as input
- discard elements from lists that are longer

so I think your implementation is fine if we stick to that 👍

Reply:
> It would also be super nice to make it work with Tuples as input, so if you do:
>
> zip(['foo', 'bar'] as const, [1, 2] as const)
> It would yield:
> ['foo', 1], ['bar', 2] on type level

I looked into this a bit, but for me it's _very hard_ to do that. The currently returned types are `["foo" | "bar", 2 | 1][]` and  I think that is good enough 👍

Reply:
Yeah, I've been banging my head against this and I can't seem to get it to infer the exact types of returned tuples.  If you do have any ideas, suggestions definitely welcome!